File: src/java/magma/app/compile/value/variant/InstanceOfs.java
package magma.app.compile.value.variant;

import magma.api.Tuple2;
import magma.api.Tuple2Impl;
import magma.app.TypeCompiler;
import magma.app.ValueCompiler;
import magma.app.compile.CompileState;
import magma.app.compile.compose.Split;
import magma.app.compile.rule.ComposableRule;
import magma.app.compile.rule.Rule;
import magma.app.compile.type.Type;
import magma.app.compile.value.Value;

public class InstanceOfs {
    public record InstanceOf(Value value, Type type) implements Value {
    }

    public static Rule<Value> createRule() {
        return new ComposableRule((CompileState state) -> Split.first(" instanceof ", (String valueString, String typeString) -> ValueCompiler.createValueRule().apply(state, valueString).flatMap(valueTuple -> {
            return TypeCompiler.createTypeRule().apply(valueTuple.left(), typeString).map((Tuple2<CompileState, Type> typeTuple) -> {
                return new Tuple2Impl<CompileState, Value>(typeTuple.left(), new InstanceOf(valueTuple.right(), typeTuple.right()));
            });
        })));
    }

    public static String generate(InstanceOf instanceOf) {
        return ValueCompiler.generateValue(instanceOf.value) + "._variant === " + TypeCompiler.generateType(instanceOf.type) + "._variantKey";
    }
}

File: src/ts/magma/app/compile/value/variant/InstanceOfs.ts
import { Value } from "../../../../../magma/app/compile/value/Value";
import { Type } from "../../../../../magma/app/compile/type/Type";
import { Rule } from "../../../../../magma/app/compile/rule/Rule";
import { Generator } from "../../../../../magma/app/compile/Generator";
export class InstanceOf implements Value {
	value: Value;
	type: Type;
	constructor (value: Value, type: Type) {
		this.value = value;
		this.type = type;
	}
}
export class InstanceOfs {
	static createRule(): Rule<Value> {
		/*return new ComposableRule((CompileState state) -> Split.first(" instanceof ", (String valueString, String typeString) -> ValueCompiler.createValueRule().apply(state, valueString).flatMap(valueTuple -> {
            return TypeCompiler.createTypeRule().apply(valueTuple.left(), typeString).map((Tuple2<CompileState, Type> typeTuple) -> {
                return new Tuple2Impl<CompileState, Value>(typeTuple.left(), new InstanceOf(valueTuple.right(), typeTuple.right()));
            });
        })))*/;
	}
	static generate(instanceOf: InstanceOf, valueGenerator: Generator<Value>, typeGenerator: Generator<Type>): string {
		return valueGenerator.generate(instanceOf.value) + "._variant === " + typeGenerator.generate(instanceOf.type) + "._variantKey"/*unknown*/;
	}
}
